airbyte-integrations/connectors/source-salesforce/source_salesforce/streams.py

```diff
@@ -607,14 +630,28 @@ def request_params(
 
         property_chunk = property_chunk or {}
 
-        stream_date = stream_state.get(self.cursor_field)
-        start_date = stream_date or self.start_date
+        start_date = max(
+            (stream_state or {}).get(self.cursor_field, self.start_date),
+            (stream_slice or {}).get("start_date", ""),
+            (next_page_token or {}).get("start_date", ""),
+        )
+        end_date = (stream_slice or {}).get("end_date", pendulum.now(tz="UTC").isoformat(timespec="milliseconds"))
+
+        select_fields = ','.join(property_chunk.keys())
+        table_name = self.name
+        where_conditions = []
+        order_by_clause = ""
 
-        query = f"SELECT {','.join(property_chunk.keys())} FROM {self.name} "
         if start_date:
-            query += f"WHERE {self.cursor_field} >= {start_date} "
+            where_conditions.append(f"{self.cursor_field} >= {start_date}")
+        if end_date:
+            where_conditions.append(f"{self.cursor_field} < {end_date}")
         if self.name not in UNSUPPORTED_FILTERING_STREAMS:
-            query += f"ORDER BY {self.cursor_field} ASC"
+            order_by_clause = f"ORDER BY {self.cursor_field} ASC"
```

Comment:
Why is it that we're only ordering by `self.cursor_field` in the incremental sync case, but ordering by cursor & primary key in the bulk case? Should they be consistent?

Reply:
We are using primary key in bulk operations query in `WHERE` clause and in `ORDER BY` clause when primary key exists. For this type of queries we need to handle pagination by ourselves so we are slicing by primary key inside base stream slicer. But rest streams pagination works in another way and we don't need it here.

**Why we need primary key for Bulk Streams:**
Given a table
|id|     date     |
|:-|:--------------|
|1|01.01.2023|
|2|01.01.2023|
|3|01.01.2023|
|4|01.01.2023|
|5|01.01.2023|
|6|01.03.2023|

Page size = 2
cursor field = date
primary key = id

Query for first slice would be:
```
SELECT fields FROM table WHERE date >= 01.01.2023 AND date < 01.02.2023 ORDER BY date LIMIT 15000;
```
Salesforce prepares data (max 15000 records but imagine it handles only 2 for example purpose):
|id|     date     |
|:-|:--------------|
|1|01.01.2023|
|2|01.01.2023|

So for now we have only 2 of 5 records satisfied first query and it means we are not ready to move to second slice. And we also see that all the 5 records have the same date `01.01.2023`. This is where `primary key` comes in handy. In order to get next two records we are making second query like that:
```
SELECT fields FROM table WHERE date >= 01.01.2023 AND date < 01.02.2023 AND id > 2 ORDER BY date, id LIMIT 15000;
```
This will return:
|id|     date     |
|:-|:--------------|
|3|01.01.2023|
|4|01.01.2023|

**Why we don't need primary key in REST Stream**:
Here (first rows of current method)
```
if next_page_token:
    """
    If `next_page_token` is set, subsequent requests use `nextRecordsUrl`, and do not include any parameters.
    """
    return {}
```
 we can see that there is a link made by Salesforce and we just use it as is for getting next page.

Reply:
Makes sense, thank you for clarifying!